### DISCUSSION TITLE
Interrupt from BAO to guests

### DISCUSSION THREAD

**Discussion by Everaldo-Gomes** [bao-project/bao-hypervisor#144] [General]
Hi, 
Can I have some directions on how to generate an interrupt in the hypervisor to be handled by the guest?

Platform: Raspberry Pi 4b
Demo: Dual-guest Linux+FreeRTOS

**Comment by josecm** [comment]
Hello Everaldo,

Simply calling `vcpu_inject_irq` and targeting the current vcpu would suffice. However, you should use an interrupt number that was not assigned to the VM in the list of devices in the configuration, as these refer to interrupts tied to an actual hardware interrupt and these should be injected using `vcpu_inject_hw_irq`. Also, you should choose an interrupt ID that makes sense for the interrupt controller in question. For the RPI4 this would mean these ID should be in the SPI spectrum. 

Let me know if this solves your problem.

**Reply by Everaldo-Gomes** [reply]
From the manual, the ID 32 should be fine.
```
vcpu_inject_hw_irq(cpu()->vcpu, 32);
vcpu_inject_irq(cpu()->vcpu, 32);
```

I just didn't understand how and where I could perform some tasks related to that CPU after calling these functions.

**Reply by josecm** [reply]
> I just didn't understand how and where I could perform some tasks related to that CPU after calling these functions.

Can you please clarify? What kinds of tasks?

**Reply by Everaldo-Gomes** [reply]
How will the guest know that it needs to handle interrupt 32 and then run a task that I created?
And how do I ensure that this will be handled by CPU 0 and not by CPU 1?

**Reply by josecm** [reply]
> How will the guest know that it needs to handle interrupt 32 and then run a task that I created?

The guest needs to be aware this interrupt exists, set it up, and act accordingly. That is up to the guet's code. For example, in our example [bare-metal application](https://github.com/bao-project/bao-baremetal-guest). The guest knows the `UART_IRQ_ID` (in your case this would be 32) and will set it up by:

1. Installing an interrupt handler which takes the required action (in this case, printing and clearing the irq):

```
void uart_rx_handler(){
    printf("cpu%d: %s\n",get_cpuid(), __func__);
    uart_clear_rxirq();
}

...

irq_set_handler(UART_IRQ_ID, uart_rx_handler);
```

2. Then, configuring the interrupt in the controller (i.e., enabling, setting priority level, etc):


```
    irq_enable(UART_IRQ_ID);
    irq_set_prio(UART_IRQ_ID, IRQ_MAX_PRIO);
```

This irq API, of course, maps to different interrupt controllers. In the case of Arm's GIC, besides enabling the interrupt it sets the calling CPU as a target for the interrupt, which should answer your last question (i.e., you program the target CPU for that interrupt in the GIC). How exactly, it really depends on the interrupt controller drivers and APIs you are using within your guest.

**Reply by Everaldo-Gomes** [reply]
It's working!
Thank you so much for your reply.
